SourceX/controls/keymapper.cpp

```diff
@@ -0,0 +1,112 @@
+#include "keymapper.hpp"
+#include "miniwin/misc.h"
+
+#include <SDL.h>
+
+#ifdef USE_SDL1
+#include "sdl2_to_1_2_backports.h"
+#endif
+
+#include <cassert>
+
+namespace devilution {
+
+Keymapper::Keymapper(SetConfigKeyFunction setKeyFunction, GetConfigKeyFunction getKeyFunction):
+    setKeyFunction{setKeyFunction},
+    getKeyFunction{getKeyFunction}
+{
+    keyIDToKeyName.reserve('Z'-'A'+'9'-'0'+12);
+    for(char c = 'A'; c <= 'Z'; ++c) {
+        keyIDToKeyName.emplace(c, std::string(1, c));
+    }
+    for(char c = '0'; c <= '9'; ++c) {
+        keyIDToKeyName.emplace(c, std::string(1, c));
+    }
+    for(int i = 0; i < 12; ++i) {
+        keyIDToKeyName.emplace(DVL_VK_F1+i, std::string(1, 'F') + std::to_string(i+1));
+    }
+
+    keyNameToKeyID.reserve(keyIDToKeyName.size());
+    for(const auto &kv: keyIDToKeyName) {
+        keyNameToKeyID.emplace(kv.second, kv.first);
+    }
+}
+
+Keymapper::ActionIndex Keymapper::addAction(const Action &action)
+{
+    actions.emplace_back(action);
+
+    return actions.size() - 1;
+}
+
+void Keymapper::keyPressed(int key, bool isPlayerDead) const
+{
+    auto it = keyIDToAction.find(key);
+    if(it == keyIDToAction.end())
+        return; // Ignore unmapped keys.
+
+    auto &action = it->second;
+
+    // Some actions cannot be triggered while the player is dead.
+    if(isPlayerDead && action.get().ifDead == Action::IfDead::Ignore)
+        return;
+
+    action();
+}
+
+std::string Keymapper::keyNameForAction(ActionIndex actionIndex) const
+{
+    assert(actionIndex < actions.size());
+    auto key = actions[actionIndex].key;
+    auto it = keyIDToKeyName.find(key);
+    assert(it != keyIDToKeyName.end());
+    return it->second;
+}
+
+void Keymapper::save() const
+{
+    // Use the action vector to go though the actions to keep the same order.
+    for(const auto &action: actions) {
```

Comment:
This has clearly not been formatted to fit the code style.

Reply:
Sure. Clang-format doesn't like braces for [initializer lists](https://www.modernescpp.com/index.php/initialization), it puts spaces before and after them. So I've used parenthesis initializer lists instead.